syntax = "proto3";

option objc_class_prefix = "RTG";

package routeguide;

service RouteGuide {
    // 最简单的rpc
    rpc GetFeature (Point) returns (Feature) {
    }
    // 服务端流式
    rpc ListFeatures (Rectangle) returns (stream Feature) {
    }
    // 客户端流式
    rpc RecordRoute (stream Point) returns (RouteSummary) {
    }
    // 请求,响应都是流式
    rpc RouteChat (stream RouteNote) returns (stream RouteNote) {
    }
}

// 圆心
// Points are represented as latitude-longitude pairs in the E7 representation
// (degrees multiplied by 10**7 and rounded to the nearest integer).
// Latitudes should be in the range +/- 90 degrees and longitude should be in
// the range +/- 180 degrees (inclusive).
message Point {
    int32 latitude = 1;
    int32 longitude = 2;
}

// 长方形,矩形
message Rectangle {
    // One corner of the rectangle.
    Point lo = 1;

    // The other corner of the rectangle.
    Point hi = 2;
}
// 面积
message Feature {
    // The name of the feature.
    string name = 1;

    // The point where the feature is detected.
    Point location = 2;
}

// 路有注释
message RouteNote {
    // The location from which the message is sent.
    Point location = 1;

    // The message to be sent.
    string message = 2;
}

message RouteSummary {
    // The number of points received. 接收的点数。
    int32 point_count = 1;

    // The number of known features passed while traversing the route. 遍历路由时传递的已知功能数。
    int32 feature_count = 2;

    // The distance covered in metres. 以米为单位的距离。
    int32 distance = 3;

    // The duration of the traversal in seconds. 遍历的持续时间（秒）。
    int32 elapsed_time = 4;
}


/*
python -m grpc_tools.protoc -I./protos --python_out=./protos --grpc_python_out=./protos ./protos/routeguide.proto

*/